packages/experiments/src/utilities/slots.tsx

```diff
@@ -73,46 +75,64 @@ export function createFactory<TProps>(
   options: IFactoryOptions<TProps> = { defaultProp: 'children' }
 ): ISlotFactory<TProps> {
   const result: ISlotFactory<TProps> = (componentProps, userProps, defaultStyles) => {
-    const propType = typeof userProps;
-
     // If they passed in raw JSX, just return that.
     if (React.isValidElement(userProps)) {
       return userProps;
     }
 
-    switch (propType) {
-      case 'string':
-      case 'number':
-      case 'boolean':
-        userProps = {
-          [options.defaultProp]: userProps as any
-        } as TProps;
-        break;
-    }
-
-    // Construct the final props for the component by merging component props, user props, and the
-    // generated class name.
-    const finalClassName = mergeStyles(defaultStyles, componentProps && componentProps.className, userProps && userProps.className);
-
-    // TODO: what if componentProps has styles prop? Here it is completely overridden by userProps without merging.
-    const finalProps = {
-      ...(componentProps as any),
-      ...(typeof userProps === 'object' && (userProps as any)),
-      className: finalClassName
-    };
-
-    // If we're rendering a function, let the user resolve how to render given the original component
-    // and final args.
+    // If we're rendering a function, let the user resolve how to render given the original component and final args.
     if (typeof userProps === 'function') {
-      return (userProps as ISlotRenderFunction<TProps>)(finalProps, ComponentType);
+      const render: ISlotRenderer<TProps> = (slotRenderFunction, renderProps) => {
+        // TODO: _translateShorthand is returning TProps, so why is the finalProps cast required?
+        // TS isn't respecting the difference between props arg type and return type and instead treating both as ISlotPropValue.
+        let finalRenderProps = _translateShorthand(options.defaultProp, renderProps) as TProps;
+        finalRenderProps = _constructFinalProps(defaultStyles, componentProps, finalRenderProps);
+
+        return slotRenderFunction(ComponentType, finalRenderProps);
+      };
+      return userProps(render);
     }
 
+    userProps = _translateShorthand(options.defaultProp, userProps);
+    // TODO: _translateShorthand is returning TProps, so why is the finalProps cast required?
+    // TS isn't respecting the difference between props arg type and return type and instead treating both as ISlotPropValue.
+    const finalProps = _constructFinalProps(defaultStyles, componentProps, userProps) as TProps;
+
     return <ComponentType {...finalProps} />;
   };
 
   return result;
 }
 
+/**
+ * Helper function that constructs a props object when given shorthand props.
+ */
+function _translateShorthand<TProps>(defaultProp: string, props: ISlotPropValue<TProps>): TProps {
+  if (typeof props === 'string' || typeof props === 'number') {
```

Comment:
you might be missing the `boolean` case

Reply:
Actually I'm going to put it back in. I guess it does make sense that a shorthand prop could be a boolean, but practically I think it will have very limited use and become confusing with logic checks against render content.